Source/AwsCommonRuntimeKit/crt/CStruct.swift

```diff
@@ -0,0 +1,41 @@
+//  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//  SPDX-License-Identifier: Apache-2.0.
+import AwsCIo
+import AwsCCommon
+// This file defines the protocols & helper functions for C Structs.
+// Instances implementing this protocol should define RawType as their C Struct.
+protocol CStruct<RawType> {
+    associatedtype RawType
+    func withCStruct<Result>(shutdownOptions: aws_shutdown_callback_options?, _ body: (RawType) -> Result) -> Result
+}
+
+extension CStruct {
+    func withCStruct<Result>( _ body: (RawType) -> Result) -> Result {
+        withCStruct(shutdownOptions: nil, body)
+    }
+
+    func withCPointer<Result>(shutdownOptions: aws_shutdown_callback_options?, _ body: (UnsafePointer<RawType>) -> Result) -> Result {
+        return withCStruct(shutdownOptions: shutdownOptions) { cStruct in
+            return withUnsafePointer(to: cStruct) { body($0) }
+        }
+    }
+
+    func withCPointer<Result>(_ body: (UnsafePointer<RawType>) -> Result) -> Result {
+        return withCStruct { cStruct in
+            return withUnsafePointer(to: cStruct) { body($0) }
+        }
+    }
+}
```

Comment:
I experimented with multiple approaches to come up with a consistent approch for all different structs. Defining a protocol helps with keep it consistent and create functions that we can apply to any C Struct like withCPointer. However, some structs had shutdown options which require special hanlding like HttpClientConnectionOptions.swift